src/many-ledger/src/storage/ledger_mintburn.rs

```diff
@@ -0,0 +1,105 @@
+use crate::error;
+use crate::storage::ledger_tokens::key_for_symbol;
+use crate::storage::{key_for_account_balance, LedgerStorage};
+use many_error::ManyError;
+use many_modules::ledger::TokenInfoArgs;
+use many_types::ledger::{LedgerTokensAddressMap, Symbol, TokenAmount, TokenInfoSupply};
+use merk::{BatchEntry, Op};
+use std::collections::BTreeSet;
+
+impl LedgerStorage {
+    pub(crate) fn get_token_supply(&self, symbol: &Symbol) -> Result<TokenInfoSupply, ManyError> {
+        Ok(self
+            .info_token(TokenInfoArgs {
+                symbol: *symbol,
+                extended_info: None,
+            })?
+            .info
+            .supply)
+    }
+
+    pub fn mint_token(
+        &mut self,
+        symbol: Symbol,
+        distribution: &LedgerTokensAddressMap,
+    ) -> Result<(), ManyError> {
+        let mut batch: Vec<BatchEntry> = Vec::new();
+        let mut circulating = TokenAmount::zero();
+        for (address, amount) in distribution.iter() {
+            circulating += amount;
+            let new_balance = self
+                .get_multiple_balances(address, &BTreeSet::from([symbol]))?
+                .get(&symbol)
+                .map_or(amount.clone(), |b| b + amount);
+            let key = key_for_account_balance(address, &symbol);
+            batch.push((key, Op::Put(new_balance.to_vec())));
+        }
+
+        // Update circulating supply
+        let mut info = self
+            .info_token(TokenInfoArgs {
+                symbol,
+                extended_info: None,
+            })?
+            .info;
+        info.supply.circulating += circulating;
+
+        // Update the supply total if the circulating supply is greater than the current total
+        if info.supply.circulating > info.supply.total {
+            info.supply.total = info.supply.circulating.clone();
+        }
+
+        batch.push((
```

Comment:
Where do you error if total > max supply?

Reply:
```rust
        // Check we don't bust the current max and that distribution doesn't contain zero
        if let Some(maximum) = current_supply.maximum {
            let amount_to_mint = distribution
                .iter()
                .fold(TokenAmount::zero(), |ref acc, (_, x)| acc + x);
            let new_amount = amount_to_mint + current_supply.circulating;
            if new_amount > maximum {
                return Err(error::over_maximum_supply(symbol, new_amount, maximum));
            }
        }
```

in `src/many-ledger/src/module/ledger_mintburn.rs`

Reply:
Yep, see above. I think we should check invariants in storage as well.